dali/kernels/imgproc/convolution/separable_convolution_cpu.h

```diff
@@ -0,0 +1,202 @@
+    req.output_shapes.push_back(uniform_list_shape<ndim>(1, in_shape));
+
+    auto req_conv = conv_.Setup(ctx, in_shape, window_sizes[0]);
+    req.AddInputSet(req_conv, false);
+
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorView<StorageCPU, Out, ndim> out,
+           const TensorView<StorageCPU, const In, ndim>& in,
+           const std::array<TensorView<StorageCPU, const W, 1>, axes>& windows,
+           const std::array<W, axes>& scales = uniform_array<axes, W>(1.f)) {
+    conv_.Run(ctx, out, in, windows[0], scales[0]);
+  }
+
+  ConvolutionCpu<Out, In, W, ndim, 0, has_channels> conv_;
+  static_assert(std::is_same<W, float>::value,
+                "Only floats as intermediate values are currently supported.");
+};
+
+template <typename Out, typename In, typename W, bool has_channels>
+struct SeparableConvolutionCpuImpl<Out, In, W, 2, has_channels> {
+  static constexpr int axes = 2;
+  static constexpr int ndim = has_channels ? 3 : 2;
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorShape<ndim>& in_shape,
+                           const std::array<int, axes>& window_sizes) {
+    KernelRequirements req;
+
+    ScratchpadEstimator se;
+    se.add<W>(AllocType::Host, volume(in_shape));
+    req.scratch_sizes = se.sizes;
+    req.output_shapes.push_back(uniform_list_shape<ndim>(1, in_shape));
+
+    auto req_inner = conv_innermost_.Setup(ctx, in_shape, window_sizes[1]);
+    auto req_outer = conv_outermost_.Setup(ctx, in_shape, window_sizes[0]);
+
+    req.AddInputSet(req_inner, false);
+    req.AddInputSet(req_outer, false);
+
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorView<StorageCPU, Out, ndim> out,
+           const TensorView<StorageCPU, const In, ndim>& in,
+           const std::array<TensorView<StorageCPU, const W, 1>, axes>& windows,
+           const std::array<W, axes>& scales = uniform_array<axes, W>(1.f)) {
+    auto *tmp = ctx.scratchpad->Allocate<W>(AllocType::Host, volume(in.shape));
+    auto intermediate = GetInterStageView(out, tmp);
+
+    conv_innermost_.Run(ctx, intermediate, in, windows[1], scales[1]);
+    conv_outermost_.Run(ctx, out, intermediate, windows[0], scales[0]);
+  }
+
+  ConvolutionCpu<W, In, W, ndim, 1, has_channels> conv_innermost_;
+  ConvolutionCpu<Out, W, W, ndim, 0, has_channels> conv_outermost_;
+  static_assert(std::is_same<W, float>::value,
+                "Only floats as intermediate values are currently supported.");
+};
+
+template <typename Out, typename In, typename W, bool has_channels>
+struct SeparableConvolutionCpuImpl<Out, In, W, 3, has_channels> {
+  static constexpr int axes = 3;
+  static constexpr int ndim = has_channels ? 4 : 3;
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorShape<ndim>& in_shape,
+                           const std::array<int, axes>& window_sizes) {
+    KernelRequirements req;
+
+    ScratchpadEstimator se;
+    se.add<W>(AllocType::Host, volume(in_shape));
```

Comment:
Likewise - you may live without this buffer.

Reply:
I know, but probably I would get a comment that it can be faster. I left it same for all variants for simplicity as well.